### ISSUE TITLE
Enhancements

### ISSUE THREAD

**Issue by PhilterPaper** [kamenov/PDF-Table#52]
I have _table_ capability on my long-term to-do list for **PDF::Builder**, but it looks like PDF::Table may do the job. There _are_ some additional features that I would like to see, that would make it far more useful, so if anyone is looking for some small projects to contribute with, these might provide some ideas:

- In HTML, the header is defined separately, defaults to bold and centered in cells, and automatically repeats. I think it should have repeat=1 by default, but since it's easy to set, it could be justified not changing. The design would have been cleaner to have a header (`<thead>`-like) defined separately from the body data, but I can live with it being the first data row.
- HTML offers a footer (`<tfoot>`), similar to the header, and PDF::Table should too. Perhaps it could be the second data line (first, if there's no header)?
- There might be a code hook to dynamically define header and footer contents (such as page subtotals).
- Either provide start-of-page and end-of-page code hooks, or make a per-page call to output that page's table. This would make it easier to put out running page headers and footers, page numbers, first and last content headings (like the first and last word on a dictionary page). Some sort of hook or flag to let the program know when the first row/column and last row/column are being processed would make it easier to output page headers.
- Offer more control over splitting the table between pages: keep the table intact unless it exceeds one page (good for short tables), allow split only only between rows (unless a row exceeds one page), and some sort of splitting within a row (at a minimum, don't create orphans and widows), allow split between paragraphs, etc. CSS might give some ideas for controls.
- API2 and Builder have a rotate(n) call to rotate the page, which can be useful for outputting wide tables. Page headers and footers should still be portrait mode, while the rest of the page content is rotated landscape. Unfortunately, I think you would have to (at page level) rotate(-90) and manually unrotate page headers and footers.
- Consider how to use PDF::Table to output continuous text in two or three columns (of one row per page), with content output (like text blocks) automatically flowing from one column to the next. The last page of a table should permit some sort of column balancing, to even out the column lengths.
- Permit vertical alignment of cell contents (see HTML and CSS valign): top, bottom, or middle.
- Define _rules_ (internal cell separators) separately from the outside enclosing _border_. Generally they would be lighter weight, or even omitted. If permitting flowed columns, allow a light vertical rule in the gutter between columns. Dashed or other pattern rules could be supported.
- Permit more elaborate border lines than just solid -- double lines, 3D beveled effects with shadows, etc. like HTML and CSS support.

I don't know when I'll have time to get to any of these, but hopefully if someone is looking for something to do, this will give some ideas (clear them with the owner first, before spending a lot of time coding!). I'll be happy to elaborate offline on what I'm looking for in a specific item, if you're not sure what I'm asking for.

**Comment by PhilterPaper** [comment]
Regarding bullet 6, I don't think `rotate()` is going to be helpful in rotating the page and table. Indeed, `rotate(-90)` _does_ rotate the page into landscape orientation, but the text is also rotated. When printed, the end result is that the rotated page is simply upside-down from the normal page.

However, it appears that calling `$page->mediabox(0,0, H,W);` (default H and W values, here reversed in order) and writing normally to it _does_ give the desired result of the page turned 90 degrees CCW and text written in that orientation, for wide tables. Page headings and footings would have to be explicitly rotated (`$text->transform(-translate=>[x,y], -rotate=>-90);`)

**Comment by PhilterPaper** [comment]
See #51 request for image support. At the moment, I'm not sure whether reserving space for an image is the best approach, or should PDF::Table be somehow rearchitected (if necessary) to allow any content at a given width and indefinite height for a cell. Having embedded images within flowed text could be quite a chore, although allowing images by themselves above or below text might not be too onerous.

**Comment by PhilterPaper** [comment]
Per #55, think about adding **row_props** to set overriding properties for an entire row. Think about expanding (generalizing) **column_props** and **cell_props** to specify "border" (actually, "rule") on 1 to 4 sides of a row, column, or cell, so effects such as a divider between two rows or two columns can be created. 

Change "border" to mean just the exterior table border, and use "rules" to talk about interior lines dividing rows, columns, and cells.

**Comment by PhilterPaper** [comment]
Update for release 1.000: `row_props` has been added. Its priority is _after_ `cell_props` and `column_props`, and before global settings. Per bullet 1, the header now repeats by default (this can be shut off).

Some sort of image support is in the wishlist/road map, for a subsequent release.

**Comment by PhilterPaper** [comment]
Items 1 and 9 done in release 1.000. Item 5 (ref #42) "to-do" list has mode to return lengths (heights) of header, footer, and each row, as well as the overall length, so you can have more control over where to break the table across pages (still between rows; no control over breaking _within_ rows). Could think about preventing widows and orphans on breaks within rows.

**Comment by PhilterPaper** [comment]
See also https://www.catskilltech.com/forum/general-12/pdftable-general-discussion/ for general discussions of direction of PDF::Table package.

**Comment by PhilterPaper** [comment]
Regarding items 3 and 4 (related), there _is_ already a `new_page_func` callback to permit you to run your own code at the _beginning_ of a page, to do things such as rotating the page and putting in _page_ headers and footers. It might be a bit difficult to do things at the _end_ of a page, unless you do them for the previous page on new_page_func invocations, and once when `table()` returns. Or, you can use the information returned by table() to go back and revisit each page, to add or revise header and footer information "after the fact" (page update). Exactly how that would be done depends on whether the page has actually been written to file and closed, or it's still in the $pdf structure. Perhaps a new `end_page_func` callback could be used for dealing with things not known until the end of the page?

Updating _table_ header row (and potentially footer row) _content_ (e.g., running totals), is an interesting problem. To go back and update the page after table() has written it means knowing exactly where the cell content is, and where you are supposed to update it _within_ that content. The `cell_render_hook` callback might help here, at least if you are willing to write the entire cell content at this time. Keep in mind that you can **not** increase the width or height of the cell (header or footer), or you'll throw all the sizing calculations off!

Finally, for item 5, one approach might be to 'virtually print' all the cells of a row (full content) and let table() decide where to split them. Or, a new setting 'heights' might prevent actually writing to the PDF, but just return an array of row heights: header, footer, row 1, row 2, ... (plus the overall single-page height). Allow 'next_y' and 'next_h' to be _arrays_ of Y-positions and heights. This way you can decide to go to the top of the next page before starting a table that could fit on one page, rather than ending up split over two pages, and control how many rows go on each page. It still wouldn't do anything about widows and orphans, or restrict where a row could be split.

**Comment by PhilterPaper** [comment]
Regarding item 5 (and #42 late request), one way to handle this (to a large degree) would be to have a setting `size` that does not output anything to the PDF file, but only returns an array of heights used. 
- **[0]** Overall height, in Points (sum of following elements, as though whole table fits on one page).
- **[1]** Height of header, 0 if not using a header.
- **[2]** Height of footer (reserved for footer), 0 if not using a footer.
- **[3]** Height of data row 1 (first).
- **[2+n]** Height of data row _n_.

This way you can look at the height and decide whether to start the table now, or jump to the top of the next page. It's especially good for short tables which can fit on one page, and you want to avoid starting them too far down the page to avoid splitting them over two pages.

For further control, you can give `next_y` and `next_h` as anonymous arrays, to control how many rows get put on each page (along with `h` for the first page). Be sure to allow for a header and/or footer on a page, as well as any caption (#31). If you run out of `next_y` and `next_h` elements, just re-use the last one as many times as necessary.

This does not correlate to any CSS settings; nor does it prevent widows and orphans if a row _does_ end up being split across two or more pages (although judicious setting of `next_h` may be helpful).

**Comment by PhilterPaper** [comment]
Since we have _colspan_, it would probably be nice to have _rowspan_ capability. Into the hopper for a near-future release.

**Comment by PhilterPaper** [comment]
Just a note so you don't overlook it: PDF::Table now supports PDF::Builder's `column()` call, which supports Markdown and HTML markup. An example is given in Builder's examples/Column.pl. This is _not_ supported in PDF::API2.

The request for "code" style display, with spaces and line-ends honored (#66), is expected to be satisfied in the near future with improvements to the `column()` call, rather than implementing them in PDF::Table itself.
